test/e2e/azure_securitygroups.go

```diff
@@ -121,6 +122,12 @@ func AzureSecurityGroupsSpec(ctx context.Context, inputGetter func() AzureSecuri
 			for _, expectedSecurityRule := range expectedSubnet.SecurityGroup.SecurityRules {
 				expectedSecurityRuleNames = append(expectedSecurityRuleNames, expectedSecurityRule.Name)
 			}
+			sort.Slice(expectedSecurityRuleNames, func(i, j int) bool {
+				return expectedSecurityRuleNames[i] < expectedSecurityRuleNames[j]
+			})
+			sort.Slice(securityRules, func(i, j int) bool {
+				return *securityRules[i].Name < *securityRules[j].Name
+			})
 
 			for _, securityRule := range securityRules {
 				g.Expect(expectedSecurityRuleNames).To(ContainElement(*securityRule.Name))
```

Comment:
I think `ContainElement` here is already protecting us from relying on a specific order for the rules.

The error isn't super clear, but I think this means that `test-security-rule` is never being deleted like we expect:
```
Expected
    <[]string | len:1, cap:1>: [
        "test-security-rule-2",
    ]
to contain element matching
    <string>: test-security-rule
```

@willie-yao do you see anything obviously wrong here?

(ref. #4264)

Reply:
I think you're right, but it would seem the flake is telling us that `test-security-rule` is actually not there (is it deleted unexpectedly?)

Reply:
I think this message is backwards w.r.t. reality if you trace through the logic of the test. Where gomega is usually `Expect(actual).To(Equal(expected))`, this is `Expect(expected).To(Equal(actual))` but I'm not sure we can just flip those here.

Either way I think this test is at least worth refactoring to make it easier to understand and that by itself might address the flake.

Reply:
> I think this message is backwards

Yup I think that if we flip it, we'll have a more accurate representation of what's going on here. We might have to compare the length of the list though to make sure there aren't any extra security rules.

Reply:
> The error isn't super clear, but I think this means that test-security-rule is never being deleted like we expect:

This is correct. It means that `test-security-rule` was in the list `securityRules` which means it wasn't deleted. Again I think flipping the for loop will let us know if the wrong rule is being deleted, or no rules are being deleted at all.

Reply:
It looks like Gomega has a built-in "check that exactly these elements exist without caring about order": https://pkg.go.dev/github.com/onsi/gomega#ContainElements

I wonder if that might more clearly express what we want here:
```diff
diff --git a/test/e2e/azure_securitygroups.go b/test/e2e/azure_securitygroups.go
index cc494b4bd..62de412a1 100644
--- a/test/e2e/azure_securitygroups.go
+++ b/test/e2e/azure_securitygroups.go
@@ -109,12 +109,14 @@ func AzureSecurityGroupsSpec(ctx context.Context, inputGetter func() AzureSecuri
                        securityGroup, err := securityGroupsClient.Get(ctx, azureCluster.Spec.ResourceGroup, expectedSubnet.SecurityGroup.Name, nil)
                        g.Expect(err).NotTo(HaveOccurred())

-                       var securityRules []*armnetwork.SecurityRule
+                       var securityRules []string
                        pager := securityRulesClient.NewListPager(azureCluster.Spec.ResourceGroup, *securityGroup.Name, nil)
                        for pager.More() {
                                nextResult, err := pager.NextPage(ctx)
                                Expect(err).NotTo(HaveOccurred())
-                               securityRules = append(securityRules, nextResult.Value...)
+                               for _, securityRule := range nextResult.Value {
+                                       securityRules = append(securityRules, *securityRule.Name)
+                               }
                        }

                        var expectedSecurityRuleNames []string
@@ -122,9 +124,7 @@ func AzureSecurityGroupsSpec(ctx context.Context, inputGetter func() AzureSecuri
                                expectedSecurityRuleNames = append(expectedSecurityRuleNames, expectedSecurityRule.Name)
                        }

-                       for _, securityRule := range securityRules {
-                               g.Expect(expectedSecurityRuleNames).To(ContainElement(*securityRule.Name))
-                       }
+                       g.Expect(securityRules).To(ConsistOf(expectedSecurityRuleNames))
                }
        }
```

Reply:
incorporated this into some more changes of mine (maybe I'm overthinking things but I didn't love trying to grok the re-use of `originalSubnets` as originally implemented as we move through the successive updates)